extensions/stats/config.cc

```diff
@@ -0,0 +1,161 @@
+  if (status != Status::OK) {
+    logWarn("cannot parse local node metadata " + node_metadata.DebugString() +
+            ": " + status.ToString());
+  }
+}
+
+inline SimpleStat resolve(bool outbound, StatGen& statgen,
+                          const common::NodeInfo& local_node_info,
+                          const common::NodeInfo& peer_node_info,
+                          const Common::RequestInfo& request_info) {
+  if (outbound) {
+    return statgen.resolve(outbound, local_node_info, peer_node_info,
+                           request_info);
+  }
+  return statgen.resolve(outbound, peer_node_info, local_node_info,
+                         request_info);
+}
+
+void PluginRootContext::report(const Common::RequestInfo& request_info) {
+  auto outbound = stats::PluginConfig_Direction_OUTBOUND == direction();
+
+  auto metadataIdKey = Common::kDownstreamMetadataIdKey;
+  auto metadataKey = Common::kDownstreamMetadataKey;
+  std::string reporter = "server";
+
+  if (outbound) {
+    metadataIdKey = Common::kUpstreamMetadataIdKey;
+    metadataKey = Common::kUpstreamMetadataKey;
+    reporter = "client";
+  }
+
+  auto peer = node_info_cache_.getPeerById(metadataIdKey, metadataKey);
+
+  // check if this peer has associated metrics
+  // These fields should vary independently of peer properties.
+  // TODO derive this from the mapper
+  auto metric_base_key =
+      absl::StrCat(peer->key, Sep, request_info.request_protocol, Sep,
+                   request_info.response_code, Sep, request_info.response_flag,
+                   Sep, request_info.mTLS);
+
+  for (auto& statgen : stats_) {
+    auto key = absl::StrCat(metric_base_key, Sep, statgen.name());
+    auto metric_it = metric_map_.find(key);
+    if (metric_it != metric_map_.end()) {
+      metric_it->second.record(request_info);
+      continue;
+    }
+
+    auto stat = resolve(outbound, statgen, local_node_info_, *peer->node_info,
+                        request_info);
+    metric_map_.insert({key, stat});
+    stat.record(request_info);
+  }
+}
+
+const NodeSharedPtr NodeInfoCache::getPeerById(StringView peer_metadata_id_key,
+                                               StringView peer_metadata_key) {
+  auto peer_id =
+      getMetadataStringValue(MetadataType::Request, peer_metadata_id_key);
+  auto nodeinfo_it = cache_.find(peer_id);
+  if (nodeinfo_it != cache_.end()) {
+    return nodeinfo_it->second;
+  }
+
+  // TODO kick out some elements from cache here if size == MAX
+
+  NodeInfoSharedPtr node_info = std::make_shared<common::NodeInfo>();
+  // Missed the cache
+  auto metadata = getMetadataStruct(MetadataType::Request, peer_metadata_key);
+  auto status = Common::extractNodeMetadata(metadata, node_info.get());
+  if (status != Status::OK) {
+    logWarn("cannot parse peer node metadata " + metadata.DebugString() + ": " +
+            status.ToString());
+  }
+
+  auto new_nodeinfo = std::make_shared<Node>(node_info);
```

Comment:
Why shared? The cache owns the node info, right? Shared pointers are generally expensive, and not advised unless required.

Reply:
If the goal is to owned by map but not shard, as we chatted, `Prefer absl::flat_hash_map<Key, std::unique_ptr<Value>> when pointer stability of values is needed.` but we can start with shard_ptr since value as unique_ptr restricted the usage of map api. This could be resolved in future PRs